File: algorithms/606. Construct String from Binary Tree/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    string tree2str(TreeNode* t) {
        if ( ! t ) return "";
        string left = t->left || t->right ? "(" + tree2str(t->left) + ")" : "";
        string right = t->right ? "(" + tree2str(t->right) + ")" : "";
        return to_string(t->val) + left + right;
    }
};

File: algorithms/606. Construct String from Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def tree2str(self, t):
        """
        :type t: TreeNode
        :rtype: str
        """
        if not t:
            return ""
        left = "(%s)" % self.tree2str(t.left) if t.left or t.right else ""
        right = "(%s)" % self.tree2str(t.right) if t.right else ""
        return "%s%s%s" % (t.val, left, right)
